File: phoneRecorder/src/main/java/com/kivsw/phonerecorder/ui/notification/NotificationShower.java
/** this class shows and hides Notofication
 *
 */
package com.kivsw.phonerecorder.ui.notification;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.BitmapFactory;
import android.support.v4.app.NotificationCompat;

import com.kivsw.phonerecorder.model.settings.ISettings;
import com.kivsw.phonerecorder.ui.main_activity.MainActivity;

import javax.inject.Inject;

import phonerecorder.kivsw.com.phonerecorder.R;

import static com.kivsw.phonerecorder.ui.notification.NotificationChannelKt.initNotificationChannel;

public class NotificationShower {

	int notificationId=0;
	protected Context context;
	ISettings settings;
	public static String CHANNEL_ID="com.kivsw.phonerecorder.notification_channel";

	//---------------------------------------------------------------------------------
	@Inject
	NotificationShower(Context context, ISettings settings, int notificationId)
	{
		this.notificationId = notificationId;
		this.context = context;
		this.settings = settings;
		initNotificationChannel(context);
	};


	/*public void setNotificationId(int notificationId) {
		this.notificationId = notificationId;
	}*/

	//---------------------------------------------------------------------------------
	public void finalize() throws Throwable
	{
		super.finalize();
		hide();
	}
	//---------------------------------------------------------------------------------
	// informs the user about this SW
	public void show(String text, boolean openActivity )
	{
		show(text, -1, openActivity);
	};

	//---------------------------------------------------------------------------------
	public void show( String text, int percents, boolean openActivity)
	{
		if(settings.getHiddenMode()) return;

		NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(context, CHANNEL_ID);
		mBuilder.setSmallIcon(R.drawable.ic_notification_small_horns);
		mBuilder.setLargeIcon( BitmapFactory.decodeResource(context.getResources(), R.drawable.ic_phonerecorder) );
		mBuilder.setContentTitle(context.getText(R.string.app_name).toString());
		mBuilder.setContentText(text);

		if(percents>=0 && percents<=100)
			mBuilder.setProgress(100, percents, false);

		if(openActivity) {
			Intent intent = new Intent(Intent.ACTION_MAIN);
			intent.setClass(context, MainActivity.class);
			intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
			intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
			mBuilder.setContentIntent(PendingIntent.getActivity(context, -1, intent, PendingIntent.FLAG_UPDATE_CURRENT));
		}

		NotificationManager mNotificationManager =
				(NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
		// notificationId allows you to update the notification later on.
		mNotificationManager.notify(notificationId, mBuilder.build());

	}
	//---------------------------------------------------------------------------------
	public void hide()
	{
		NotificationManager mNotificationManager =
				(NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
		mNotificationManager.cancel(notificationId);
	}

}

File: phoneRecorder/src/main/java/com/kivsw/phonerecorder/ui/notification/NotificationShower.kt
/** this class shows and hides Notofication
 *
 */
package com.kivsw.phonerecorder.ui.notification

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.support.v4.app.NotificationCompat
import com.kivsw.phonerecorder.model.settings.ISettings
import com.kivsw.phonerecorder.ui.main_activity.MainActivity
import phonerecorder.kivsw.com.phonerecorder.R
import javax.inject.Inject

open class NotificationShower
@Inject
internal constructor(protected var context: Context, internal var settings: ISettings, internal var notificationId: Int) {

    init {
        initNotificationChannel(context)
    }

    //---------------------------------------------------------------------------------
    //https://kotlinlang.org/docs/reference/java-interop.html#finalize
    @Throws(Throwable::class)
    fun finalize() {
        //super.finalize() // no needs to invoke it in Kotlin
        hide()
    }

    //---------------------------------------------------------------------------------
    fun createNotification(title:String, text: String, iconId:Int, largeIcon: Bitmap,
                                     percent:Int=-1, pendingIntent:PendingIntent?=null ): Notification
    {
        val mBuilder = NotificationCompat.Builder(context, CHANNEL_ID)
        mBuilder.setSmallIcon(iconId)
        mBuilder.setLargeIcon(largeIcon)
        mBuilder.setContentTitle(title)
        mBuilder.setContentText(text)
        mBuilder.setOngoing(true)

        if (percent >= 0 && percent <= 100)
            mBuilder.setProgress(100, percent, false)

        pendingIntent?.let{
            mBuilder.setContentIntent(pendingIntent)
        }

        return mBuilder.build()
    }
    //---------------------------------------------------------------------------------
    // informs the user about this SW
    fun show(text: String, openActivity: Boolean) {
        show(text, -1, openActivity)
    }

    //---------------------------------------------------------------------------------
    fun show(text: String, percents: Int, openActivity: Boolean) {
        if (settings.hiddenMode) return

        val intent = Intent(Intent.ACTION_MAIN)
            intent.setClass(context, MainActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)

        val n=createNotification(context.getText(R.string.app_name).toString(),
                text,
                R.drawable.ic_notification_small_horns,
                BitmapFactory.decodeResource(context.resources, R.drawable.ic_phonerecorder),
                percents,
                PendingIntent.getActivity(context, -1, intent, PendingIntent.FLAG_UPDATE_CURRENT))

        val mNotificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        // notificationId allows you to update the notification later on.
        mNotificationManager.notify(notificationId, n)
    }

    //---------------------------------------------------------------------------------
    fun hide() {
        val mNotificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        mNotificationManager.cancel(notificationId)
    }

    companion object {
        var CHANNEL_ID = "com.kivsw.phonerecorder.notification_channel"
    }

}

